core/src/backend/video/software.rs

```diff
@@ -198,3 +202,194 @@ mod h263 {
         }
     }
 }
+
+#[cfg(feature = "vp6")]
+mod vp6 {
+    use crate::backend::video::software::VideoDecoder;
+    use crate::backend::video::{DecodedFrame, EncodedFrame, Error, FrameDependency};
+
+    use h263_rs_yuv::bt601::yuv420_to_rgba;
+
+    use nihav_codec_support::codecs::NAVideoInfo;
+    use nihav_codec_support::codecs::{NABufferType::Video, YUV420_FORMAT};
+    use nihav_core::codecs::NADecoderSupport;
+    use nihav_duck::codecs::vp6::{VP56Decoder, VP56Parser, VP6BR};
+    use nihav_duck::codecs::vpcommon::{BoolCoder, VP_YUVA420_FORMAT};
+
+    /// VP6 video decoder.
+    pub struct Vp6Decoder {
+        with_alpha: bool,
+        bounds: (u16, u16),
+        decoder: VP56Decoder,
+        support: NADecoderSupport,
+        bitreader: VP6BR,
+        init_called: bool,
+    }
+
+    impl Vp6Decoder {
+        pub fn new(with_alpha: bool, bounds: (u16, u16)) -> Self {
+            // Unfortunately, `init()` cannot be called on the decoder
+            // just yet, because `bounds` is only the declared size of
+            // the video, to which it will be cropped.
+            // This can be (rarely) even much smaller than the actual
+            // encoded size of the frames.
+            // `VP56Decoder::init()` needs the full encoded frame size,
+            // so it can allocate its internal buffers accordingly.
+            // The encoded frame size will be parsed from the header of
+            // the first encoded frame passed to `Self::decode_frame()`.
+
+            Self {
+                with_alpha,
+                bounds,
+                decoder: VP56Decoder::new(6, with_alpha, true),
+                support: NADecoderSupport::new(),
+                bitreader: VP6BR::new(),
+                init_called: false,
+            }
+        }
+    }
+
+    impl VideoDecoder for Vp6Decoder {
+        fn preload_frame(
+            &mut self,
+            encoded_frame: EncodedFrame<'_>,
+        ) -> Result<FrameDependency, Error> {
+            // Luckily the very first bit of the encoded frames is exactly this flag,
+            // so we don't have to bother asking any "proper" decoder or parser.
+            Ok(
+                if !encoded_frame.data.is_empty() && (encoded_frame.data[0] & 0b_1000_0000) == 0 {
+                    FrameDependency::None
+                } else {
+                    FrameDependency::Past
+                },
+            )
+        }
+
+        fn decode_frame(&mut self, encoded_frame: EncodedFrame<'_>) -> Result<DecodedFrame, Error> {
+            // If this is the first frame, the decoder needs to be initialized.
+
+            if !self.init_called {
+                let mut bool_coder = BoolCoder::new(encoded_frame.data).unwrap();
+                let header = self.bitreader.parse_header(&mut bool_coder).unwrap();
```

Comment:
Can we return the error here instead of `unwrap`?

Reply:
I changed the `unwrap()`s to `.map_err(...)?` patterns.